executor/executor.cc

```diff
@@ -519,8 +535,21 @@ static execute_req last_execute_req;
 void receive_execute()
 {
 	execute_req& req = last_execute_req;
+	uint64_t req_pos = 0;
+	#if GOOS_windows
+	for (;;)
+	{
+		signed int rv = read(kInPipeFd, &req, sizeof(req));
+		if(rv < 0)
+			fail("control pipe read failed receive_execute");
+		req_pos += rv;
```

Comment:
I don't understand this part.
I can imagine read from the pipe is split into parts, but you always start filling req from the start, not from req_pos offset. This means that if the code works for you, we don't need this change and the pipe never returns incomplete data.

Reply:
Okay I'm reverting it back to normal. This is the part where the current executor fails.

Reply:
Okay so going back to the way it was fixes the broken pipe. The only reason the current build of executor fails on Windows is because of the way the input is read
```
bad input size 656, want 4920 (errno 0) 
executor 0: got bad reply magic 0xadf00d00
failed to execute executor: executor 2: exit status 67 
bad input size 974, want 7944 (errno 0)
```

Reply:
This is one is a different place:
> bad input size 656, want 4920 (errno 0) 
where we already have read loop that reads in parts, right?

Does it happen all the time? Or sometimes?
The output you posted: is it consecutive lines of the same log? Or cherry-picked lines from 4 different logs?

Reply:
The output is from a single run. It would be better to look at the output while running Syz-stress with procs=1.
```
2020/07/23 01:21:33 parsed 0 programs
2020/07/23 01:21:33 unsupported syscall: syz_builtin0: has disabled attribute in descriptions
2020/07/23 01:21:33 unsupported syscall: syz_builtin1: has disabled attribute in descriptions
2020/07/23 01:21:33 unsupported syscall: syz_execute_func: always disabled for now
failed to execute executor: executor 0: exit status 67
bad input size 700, want 4120 (errno 0)
------PROGRAM REDACTED-----
bad input size 700, want 4120 (errno 0)
failed to execute executor: executor 0: exit status 67
bad input size 700, want 4120 (errno 0)
------PROGRAM REDACTED-----
bad input size 700, want 4120 (errno 0)
executor 0: got bad reply magic 0xadf00d00
```
Syz-stress ends once it receives the bad reply magic.

Reply:
So somehow the read from the pipe returns 0 in the middle of the message....
In the last log all 3 failures have number 700 which looks somewhat specious and not too random. Is the number 700 involved somewhere else?
Do you have any other changes? Perhaps there is a bug on the other side of the pipe? How many bytes are written to the pipe?

Reply:
Syz-stress was built from the same branch. So I don't have any other changes made.

Reply:
Humm... then I am out of ideas.
I would log what's exactly being written to the pipe and in what portions, what's being read, in what portions, what's the status of the pipe after the failure, etc. That may shed some light.

Your original change helped with this? It was in the beginning of the function where we read fixed-size header, right? I don't see how it can help the failures while reading the variable-size portion... do you have any hypothesis?

Reply:
No I don't think my change helped either. I just copied the read part form the bottom and put it there to check if that would help. I'm currently debugging it and looks like the pipe is writing a lot of ```fe ff ff ff ff ff ff ff```

The memory dump of first 128 bytes is as follows.
```
00007ff6`91337e40  fe ff ff ff ff ff ff ff-00 00 00 20 00 00 00 00 
00007ff6`91337e50  00 00 00 00 00 00 00 00-02 00 00 00 00 00 00 00 
00007ff6`91337e60  01 00 00 00 00 00 00 00-fe ff ff ff ff ff ff ff 
00007ff6`91337e70  40 00 00 20 00 00 00 00-00 00 00 00 00 00 00 00 
00007ff6`91337e80  01 00 00 00 00 00 00 00-43 00 00 00 00 00 00 00 
00007ff6`91337e90  fe ff ff ff ff ff ff ff-80 00 00 20 00 00 00 00 
00007ff6`91337ea0  00 00 00 00 00 00 00 00-08 00 00 00 00 00 00 00 
00007ff6`91337eb0  07 00 00 00 00 00 00 00-fe ff ff ff ff ff ff ff
```
Looking back at how the initial build of Windows. This was written to STDIN and STDOUT.

Reply:
Looking at the implementation of pipes in Linux and Windows one major thing that sets them both apart is the way you can access them. In Linux you access the pipes using descriptors but in Windows these have to be handles. Does this issue have something in common with this [os: Pipe can't be inherited on windows](https://github.com/golang/go/issues/21085) issue in golang. We could have a different file called ipc_windows.go which follows the previous implementation of using STDIN,STDERR and STDOUT.

Reply:
If we don't inherit them, I would assume the very first read would fail with some clear error. But read something from it, which means we inherit it, right.

Reply:
Oh okay I debug and try figuring something out for this.

Reply:
Hi @dvyukov So the issue was in the second wherein we read the progData. Some of the reads return 0 which I'm not quite sure on why it happens. Removing ```rv ==0`` from line [577](https://github.com/google/syzkaller/blob/cb93dc6ac64225e09f44bac6c6cce1dae1b248b0/executor/executor.cc#L557) made the executor read the whole program and execute it.
![image](https://user-images.githubusercontent.com/16324259/88649422-7824fe80-d0e5-11ea-9199-01eccfdab870.png)
Here is a screenshot of executor running a test of GetProcessId(). I'm not quite sure if this is a right thing to do or not.

Reply:
You mean read returns 0, but the next time it returns a non-value? Well, it would be a kernel bug for any unix I think.
What does MSDN say for 0 return values for the read? Do we call write with 0 size on the pipe?

Reply:
The read returns 0 quite a number of times. Looking at MSDN it says that it happens when the pipe is empty and it doesn't have anything to read. No we don't call it with size 0 . We write the progData.

Reply:
This means we have pipe in nonblocking-wait mode:
https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipe-type-read-and-wait-modes#wait-mode
We should try to switch it to blocking-wait mode. Please check if we are providing correct flags to CreateNamedPipe/CreateFile or whatever syscall we use. Or if it's not possible to fix them, add SetNamedPipeHandleState to os_init.

Reply:
But it looks like Golang doesn't support SetNonBlock on Windows. https://golang.org/src/os/pipe_test.go#L208

Reply:
I am not sure this is relevant.
Go supports everything. It can call arbitrary system calls with arbitrary arguments.
Also I think SetNonblock is used to switch files to non-blocking, they should be blocking by default.
We should create the pipe as blocking.

Reply:
Okay the best thing to do was use ReadFile and WriteFile when it's Windows. ```SetNamedPipeHandleState``` did not help much as the write was failing due to the wait. By using ReadFile and WriteFile there is no need of a loop since it does everything on it own. The down side to this is another if else preprocessor block is added.